include/cgimap/http.hpp

```diff
@@ -115,17 +116,32 @@ class gone : public exception {
 };
 
 /**
+ * Indicates that the client is not authorized to perform the request.
+ */
+class unauthorized : public exception {
+public:
+  unauthorized(const std::string &message);
+};
+
+/**
  * Decodes a url-encoded string.
  */
 std::string urldecode(const std::string &s);
 
 /**
+ * Encodes a string to canonical url-encoding.
+ *
+ * This is compatible with the OAuth 1.0a definition of url-encoding.
+ */
+std::string urlencode(const std::string &s);
+
+/**
  * Parses a query string into a key-value map.
  *
  * The string should already have been url-decoded (i.e: no %-encoded
  * chars remain).
  */
-std::map<std::string, std::string> parse_params(const std::string &p);
+std::vector<std::pair<std::string, std::string> > parse_params(const std::string &p);
```

Comment:
What's the reason for changing from a map? If it's to allow duplicates is multimap not the way to go?